src/cache.rs

```diff
@@ -0,0 +1,283 @@
+        self.cache.contains_key(key)
+    }
+
+    fn add<'sqlite, 'schema>(&mut self,
+                                            sqlite: &'sqlite rusqlite::Connection,
+                                            schema: &'schema Schema,
+                                            key: Self::Key) -> Result<()> {
+        // fetch results and add to cache
+        let values = self.value_provider.values_for_attribute(sqlite, schema, &key)?;
+        self.cache.insert(key.clone(), values);
+        Ok(())
+    }
+
+    fn remove(&mut self, key: &Self::Key) -> Result<Self::Value> {
+        let r = self.cache.remove(key).ok_or_else(||ErrorKind::CacheMiss(format!("{:?}", key)))?;
+        Ok(r)
+    }
+
+    fn get(&mut self, key: &Self::Key) -> Result<&Self::Value> {
+        Ok(self.cache.get(&key).ok_or_else(|| ErrorKind::CacheMiss(format!("{:?}", key)))?)
+    }
+}
+
+pub struct AttributeCache {
+    cache: EagerCache<NamespacedKeyword, BTreeMap<TypedValue, TypedValue>, AttributeValueProvider>,   // values keyed by attribute
+}
+
+impl AttributeCache {
+
+    pub fn new() -> Self {
+        AttributeCache {
+            cache: EagerCache::new(AttributeValueProvider::default()),
+        }
+    }
+
+    fn contains_key(&self, key: &NamespacedKeyword) -> bool {
+        self.cache.contains_key(key)
+    }
+
+    pub fn add_to_cache<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema, key: NamespacedKeyword) -> Result<()> {
+        self.cache.add( sqlite, schema, key)
+    }
+
+    pub fn remove_from_cache(&mut self, key: &NamespacedKeyword) -> Result<BTreeMap<TypedValue, TypedValue>> {
+        self.cache.remove(key)
+    }
+
+    pub fn get<'sqlite, 'schema>(&mut self, key: &NamespacedKeyword) -> Result<&BTreeMap<TypedValue, TypedValue>> {
+        self.cache.get( key)
+    }
+
+    pub fn get_for_entid(&mut self, key: NamespacedKeyword, entid: KnownEntid) -> Result<Option<&TypedValue>> {
+        let entity: TypedValue = entid.into();
+        let map = self.cache.get(&key)?;
+        Ok(map.get(&entity))
+    }
+}
+
+#[derive(Default)]
+struct AttributeValueProvider{}
+
+impl ValueProvider<NamespacedKeyword, BTreeMap<TypedValue, TypedValue>> for AttributeValueProvider {
+    fn values_for_attribute<'sqlite, 'schema>(&mut self, sqlite: &'sqlite rusqlite::Connection, schema: &'schema Schema, attribute: &NamespacedKeyword) -> Result<BTreeMap<TypedValue, TypedValue>> {
+        let entid = schema.get_entid(&attribute).ok_or_else(|| ErrorKind::UnknownAttribute(format!("{:?}", attribute)))?;
+        Ok(q_once(sqlite,
+                  schema,
+                  r#"[:find ?entity ?value
+                                          :in ?attribute
+                                          :where [?entity ?attribute ?value]]"#,
+                  QueryInputs::with_value_sequence(vec![(Variable::from_valid_name("?attribute"), TypedValue::Long(entid.0))]))
+            .into_rel_result()?
+            .iter()
```

Comment:
`into_iter()`, then you don't need to clone.

Reply:
I clearly don't know how to use `into_iter` properly as even using that I still needed to clone...